### PR TITLE
Add support for Doxygen pages

### LINKED ISSUE #111: [warning] Unabridged API: unexpected kind 'page' (IGNORED)
## What
Doxygen is producing `page` tags for **markdown** files that are part of the `include` list in my project's Doxyfile. Currently, `exhale` is throwing a warning that reads `(!) Unabridged API: unexpected kind 'page' (IGNORED)`. This is not a *huge* problem, because `lexerMapping` is supported, but it would be nice to get over this warning - both to have stable CI on my project, and to have builtin support for markdown pages.

## Why
[It looks like `exhale` does not qualify `page` as an input kind](https://exhale.readthedocs.io/en/latest/reference/utils.html#exhale.utils.AVAILABLE_KINDS), and hence the appropriate defaults are not used for this input kind. I think a simple fix would be to allow users to append to `exhale.utils.AVAILABLE_KINDS` through `exhale_args`, and/or add support for the `page` input kind, which would map to a `doxygenpage` output directive for `breathe`. For the former option, users can do their own `"customSpecificationsMapping"` thing, and for the latter option, since there is a 1-to-1 output directive mapping, support for pages could be formally introduced within `graph.py`.

----

@svenevs: I would be happy to contribute through the second suggestion if needed. I would need some guidance though because `graph.py` is ginormous and I couldn't really make of where such a patch has to go.

Also, thanks a lot for the thorough documentation throughout the project!

### PR DESCRIPTION
**Pull request by hidmic**
Closes #111. This patch is a first stab at supporting Doxygen pages. Namely, it:

- Includes Doxygen's main page in the root file
- Allows page listings in the unabridged API
- Displays a page hierarchy (ignoring the main page)
- Makes the root title optional to avoid duplicates

This is lacking tests (and some existing tests are failing on `master`), but I'd like to get some preliminary feedback.

### CODE REVIEW

**Comment by svenevs** [comment]
> preliminary feedback

Thanks for putting out a PR on this!  Obviously, I've been neglecting this repo... as you probably discovered, there's a lot of technical debt here, a lot mine, and a lot not mine.  I've fixed up most of the CI in #115 but need to switch over to GitHub Actions and ditch Travis.  I'm moving soon but have some time to squeak out a release and get pages working.  This one should have tests added, but I can help with that (the testing framework is a lot).

To be honest, I've never really used doxygen pages.  I'll try and build your linked PR and see what's going on, initially I'm apprehensive to remove `rootFileTitle` as a required argument.  See https://exhale.readthedocs.io/en/latest/usage.html#additional-usage-and-customization

1. Can you update that table (`docs/reference/configs.rst`, search `begin_root_api_document_layout`) to include where the pages go?
2. Linked https://exhale.readthedocs.io/en/latest/reference/configs.html#exhale.configs.rootFileTitle docs, sphinx uses the `rootFileTitle` wherever the `.. toctree::` that includes it is.  I'm not seeing why this can be relaxed in the presence of pages, but perhaps that will become more clear after I build it.

File: exhale/graph.py [hunk 1]
```diff
@@ -564,6 +565,19 @@ def typeSort(self):
...
+        '''
+        Whether or not this node should be included in the page view hierarchy.  Helper
+        method for :func:`~exhale.graph.ExhaleNode.toHierarchy`.  Sets the member
+        variable ``self.in_page_hierarchy`` to True if appropriate.
+
+        :Return (bool):
+            True if this node should be included in the page view --- if it is a
+            node of kind ``page``. Returns False otherwise.
+        '''
+        self.in_page_hierarchy = (self.kind == "page")
```
Comment (svenevs) [inline exhale/graph.py:578]:
```diff
- self.in_page_hierarchy = (self.kind == "page")
+ self.in_page_hierarchy = self.kind == "page"
```

parens not needed.  from what I'm gathering, pages don't have any hierarchy?  seems legit

File: exhale/graph.py [hunk 2]
```diff
@@ -615,14 +629,24 @@ def inFileHierarchy(self):
                     return True
         return False
 
-    def inHierarchy(self, classView):
-        if classView:
+    def inHierarchy(self, classType):
```
Comment (svenevs) [inline exhale/graph.py:632]: Lets go ahead and rename this to just `hierarchy` or `hierarchyType` for the second parameter (dealers choice).  Probably make this an enum once 2.7 is ditched, but for now we'll work with strings :slightly_smiling_face:

File: exhale/graph.py [hunk 3]
```diff
@@ -744,20 +771,26 @@ def toHierarchy(self, classView, level, stream, lastChild=False):
...
                 else:
-                    # E.g. 'Class Foo'
-                    q_start = 0
-                    q_end   = 1
-                # the qualifier will not be part of the hyperlink (for clarity of
-                # navigation), the link_title will be
-                qualifier   = " ".join(title_as_link_parts[q_start:q_end])
-                link_title  = " ".join(title_as_link_parts[q_end:])
+                    # E.g. 'Foo'
+                    qualifier = ""
```
Comment (svenevs) [inline exhale/graph.py:791]: Nice :boom:

File: exhale/graph.py [hunk 4]
```diff
@@ -966,6 +999,7 @@ def __init__(self):
         self.root_directory         = configs.containmentFolder
         self.root_file_name         = configs.rootFileName
         self.full_root_file_path    = os.path.join(self.root_directory, self.root_file_name)
+        self.page_hierarchy_file    = os.path.join(self.root_directory, "page_view_hierarchy.rst")
```
Comment (svenevs) [inline exhale/graph.py:1002]: In docs request for table update on `begin_root_api_document_layout` in `docs/reference/configs.rst` please explain this is the file.  I've seen people manually post-process these files, so it's nice to know what it is.

File: exhale/graph.py [hunk 5]
```diff
@@ -1116,6 +1152,75 @@ def discoverAllNodes(self):
...
+                        ))
+                        if location_str.startswith(abs_strip_path):
+                            location_str = os.path.relpath(location_str, abs_strip_path)
+                        page.location = os.path.normpath(location_str)
+
+                except:
+                    utils.fancyError(
+                        "Could not process Doxygen xml for file [{0}]".format(f.name)
+                    )
+        self.pages = [page for page in self.pages if not page.parent]
```
Comment (svenevs) [inline exhale/graph.py:1222]: Ok so basically you avoid needing to reparent by just setting the children as you set this up, and we can rely on innerpage definitions being present?  Nice

File: exhale/graph.py [hunk 6]
```diff
@@ -2655,6 +2769,82 @@ def generateSingleNodeRST(self, node):
...
+        '''
+        Generates the reStructuredText document for every page.
+        '''
+        for page in self.pages:
+            # generate this page
+            self.generateSinglePageDocument(page)
+
+    def generateSinglePageDocument(self, node):
+        '''
+        Creates the reStructuredText document for a page
```
Comment (svenevs) [inline exhale/graph.py:2782]: Add period at end of sentence please :slightly_smiling_face:

File: exhale/graph.py [hunk 7]
```diff
@@ -3387,20 +3602,42 @@ def gerrymanderNodeFilenames(self):
...
         :func:`~exhale.graph.ExhaleRoot.generateAPIRootBody`.
         '''
+        # gather the page hierarchy data and write it out
+        page_view_data = self.generatePageView()
+        self.writeOutHierarchy({
+            "idx": configs._page_hierarchy_id,
+            "bstrap_data_func_name": configs._bstrap_page_hierarchy_fn_data_name,
+            "file_name": self.page_hierarchy_file,
+            "file_title": "Content",
+            "type": "page"
```
Comment (svenevs) [inline exhale/graph.py:3619]: Whatever the change is above (`hierarchy` vs `hierarchyType`), lets make `writeOutHierarchy` expect the same thing.

**Review by svenevs — CHANGES_REQUESTED** [review]
Overall, it looks great, see above comment will check it out more closely soon!

From what I'm gathering here, `if p.refid == "indexpage":` you're wanting to remove the `rootFileTitle` because of this one.  If you could orient me a bit on the tension points, and more generally how peoples use doxygen pages, I'm sure we can find a compromise or hack.  The error checking was a little overly obsessive, it seems kind of like we want a switch

- If project has `page` entries: don't require `rootFileTitle`
- Else: require it (sphinx requires you to have a title)

So leaning toward keep it required, and just explicitly say it will not get used if you are also using pages?

**Comment by svenevs** [comment]
Ok quick update, CI was fixed, will be able to look more closely at the implementation here over the next couple of days / get acquainted with doxygen pages.

**Comment by svenevs** [comment]
Ok I added dracula in with some doxygen page stuff, no actual testing done right now, there's some issues with the current implementation and the rst links.

```console
$ cd /path/to/exhale
# (or however you manage py installs, get tox)
$ pip install tox

# Run just the tests associated with the project that has the doxygenpage examples
$ tox -e py -- -k cpp_nesting -s

# See the issues in current html output
$ find testing/projects/cpp_nesting -name index.html -exec xdg-open {} \;

# (alternatively open them manually)
$ find testing/projects/cpp_nesting -name index.html
testing/projects/cpp_nesting/docs_CPPNestingPages_test_hierarchies_with_pages_dracula_not_mainpage/_build/html/index.html
testing/projects/cpp_nesting/docs_CPPNestingPages_test_hierarchies_with_pages_dracula_is_mainpage/_build/html/index.html
```

Action items:

- [ ] Do not force the word `Contents` on people, that can be a configuration variable.
- [ ] Fix the rst links (note tree view is not enabled in those builds)
    - [ ] AFAICT the `page` kind of doxygen in xml will not have a `def_in_file`, we can omit those warnings and comment.  That may be the break.
- [ ] Validate tree view manually (by inspection)
- [ ] Actually validate tests.

What I'm not understanding:

1. Why is there any parent relationship stuff going on?  I don't understand how nested pages come in or how to create them.
2. Seems to be related to you removing the title.  Why does the `\mainpage` version remove the Class / File hierarchy from the `toctree`?  I don't like this behavior, can it be opt in?
3. Do you have a sample project I can look at (rather than just pasting Dracula in...) so I can actually see what you're trying to do?  That'll probably make things clearer :slightly_smiling_face:

Mostly just taking notes to myself here, but will get back at it soon.

**Comment by hidmic** [comment]
> From what I'm gathering here, `if p.refid == "indexpage"`: you're wanting to remove the `rootFileTitle` because of this one. If you could orient me a bit on the tension points, and more generally how peoples use doxygen pages, I'm sure we can find a compromise or hack.

Right. The basic idea is that if your documentation is already organized in pages within Doxygen, there's no need for `rootFileTitle`. `\mainpage` will already have a title.

> So leaning toward keep it required, and just explicitly say it will not get used if you are also using pages?

I don't have a strong opinion here but I will point out that Doxygen may have pages but no main page, in which case you do want a `rootFileTitle`. Only a user knows whether they need it or not. Thus why I made it optional. 

> Why is there any parent relationship stuff going on? I don't understand how nested pages come in or how to create them.

Doxygen supports page hierarchies via [`\subpage`](https://www.doxygen.nl/manual/commands.html#cmdsubpage) refs. 

> Seems to be related to you removing the title. Why does the \mainpage version remove the Class / File hierarchy from the toctree? I don't like this behavior, can it be opt in?

Hmm, I suspect this is related to `writeOutHierarchy` changes. I changed it so that it won't write out a hierarchy if it is empty (e.g. no class hierarchy if there's no API to show).

> Do you have a sample project I can look at (rather than just pasting Dracula in...) so I can actually see what you're trying to do? That'll probably make things clearer 

Sure. There's changes in flight so it's not super straightforward to reproduce. Check [this bash gist](https://gist.github.com/hidmic/ad6c00a88bd2975da1e0775744da739b).

File: exhale/graph.py (hunk 5 above)
Comment (hidmic) [reply exhale/graph.py:1222]: Exactly. Checking Doxygen generated XML, I believe this holds.

File: exhale/graph.py (hunk 3 above)
Comment (hidmic) [reply exhale/graph.py:791]: Haha, yeah. I didn't put a lot of effort in trying to generalize this bit. I just find having a leading `Page` noun somewhat redundant.

**Comment by hidmic** [comment]
Thanks for reviewing @svenevs! I agree with most of your points. I'll try to address your comments sooner than later.

File: exhale/graph.py (hunk 7 above)
Comment (hidmic) [reply exhale/graph.py:3633]: Hmm, do you think that `hierarchy_config['type']` isn't explicit enough?

Comment (svenevs) [reply exhale/graph.py:3633]: Oh yeah that's fine as is, sorry!

File: exhale/graph.py (hunk 2 above)
Comment (hidmic) [reply exhale/graph.py]: Done in https://github.com/svenevs/exhale/pull/114/commits/674aa4be1f70b79560f741df18599da7d715fa36.

File: exhale/graph.py (hunk 4 above)
Comment (hidmic) [reply exhale/graph.py:1002]: Done in https://github.com/svenevs/exhale/pull/114/commits/c4b141cc1fc852df08d74019a5dcb8cc3e2aa8c3

**Comment by hidmic** [comment]
>  Do not force the word Contents on people, that can be a configuration variable.

Done in https://github.com/svenevs/exhale/pull/114/commits/72b5c4de0d8ebb98e4f3735970e6b73d032765b0.

> If I were you I'd hang on a bit before jumping back in on this, the testing machinery is not something you should have to deal with.

Got it. I'll stop here then.

**Comment by 2bndy5** [comment]
<details><summary>Here's a modified example snippet from the doxygen manual</summary>

```cpp
/** \mainpage A simple manual
 *
 * Some general info.
 *
 * \section subpage_list A list of subpages
 * This manual is divided in the following sections:
 * - \subpage intro
 * - \subpage advanced "Advanced usage"
 *
 * \section non-op
 * This section has no explicit title.
 *
 * \page intro Introduction
 * This page introduces the user to the topic.
 * Now you can proceed to the \ref advanced "advanced section".
 *
 * \section howto Tutorial
 *
 * \subsection getting_started Getting Started
 * Just kidding. Go fish!
 *
 * \page advanced Advanced Usage
 * This page is for advanced users.
 * Make sure you have first read \ref intro "the introduction".
 *
 * \section pre-req Things to know first
 *
 * \subsection fair_warning Here Be Dragons
 * Best to avoid dragons.
 *
 * \subsubsection extinction Dragon's are a myth?!
 * There's an interesting history to the phrase \ref fair_warning "'Here Be Dragons'"
 */
```
</details>

<details><summary> here is a sample MD file syntax</summary>

```md
# h1 Title
<!-- markdownlint-disable -->
This section's title gets used as the page title.
Thus, the heading isn't rendered in breathe, but the section contents are.
If the first heading is not a h1 syntax, then the MD file's name is used.

@warning
If a MD comment exists before the first h1 heading,
then the title of the page defaults the MD file's name.
In that case, the h1 heading is rendered as a `@section`.

## h2 List syntax gotchas
1. Item1 of list 1
3. Item2 of list 1
2. Item1 of list 2
4. Item2 of list 2

### h3 comparing to section CMD
This is section is similar to doxygen's `@subsection` command.

#### h4 Yet another nested section
This is section is similar to doxygen's `@subsubsection` command.

##### h5 tricks of trade
There seems to be no equivalent doxygen command for h5 level sections.
This section is not actually rendered with breathe.

@note This h5 level heading is the most subsections deep that get shown in TREEVIEW

###### h6 The limit of MD headings
According to Markdown syntax specs, there is no way to go beyond h6 level headings.
This section is not actually rendered with breathe.
```

File: testing/tests/cpp_nesting.py [hunk 8]
```diff
@@ -26,114 +41,128 @@ class CPPNesting(ExhaleTestCase):
...
+        ``page_hierarchy``.
+
+    ``page_hierarchy`` (:class:`python:dict`)
+        A dictionary of :ref:`~testing.hierarchies.mainpage`,
+        :ref:`~testing.hierarchies.page`, and/or :ref:`~testing.hierarchies.subpage`
+        keys that either map to a :ref:`~testing.hierarchies.page_contents` node, or
+        a subpage.
+
+    **Return**
+    The input ``func``, after executing ``pytest.mark.usefixtures``.
```
Comment (2bndy5) [inline testing/tests/cpp_nesting.py:84]: This docstr is throwing "unexpected indent" errors in sphinx. It should use sphinx compatible rst syntax.
```suggestion
    :param str f_path: The destination to generate the C++ documented file encoding the
        specified ``page_hierarchy``.
    :param dict page_hierarchy: A dictionary of :ref:`~testing.hierarchies.mainpage`,
        :ref:`~testing.hierarchies.page`, and/or :ref:`~testing.hierarchies.subpage`
        keys that either map to a :ref:`~testing.hierarchies.page_contents` node, or
        a subpage.

    :Returns: The input ``func``, after executing ``pytest.mark.usefixtures``.
```

File: testing/tests/cpp_nesting.py [hunk 9]
```diff
@@ -11,13 +11,28 @@
...
+
+import pytest
+
+from testing import TEST_PROJECTS_ROOT
 from testing.base import ExhaleTestCase
-from testing.decorators import confoverrides
-from testing.hierarchies import                                              \
-    class_hierarchy, clike, compare_class_hierarchy, compare_file_hierarchy, \
-    directory, file, file_hierarchy, namespace, union
+from testing.decorators import confoverrides, no_cleanup
```
Comment (2bndy5) [inline testing/tests/cpp_nesting.py:22]: not sure if this was a copy-n-paste problem, but flake8 is having a problem with the unused import `no_cleanup`

```suggestion
from testing.decorators import confoverrides
```

**Review by 2bndy5 — COMMENTED** [review]
I can't submit changes because of obvious lack of permission, so I left some comments with suggestions that should fix the "Test Extras" CI jobs.

I don't really know what's going wrong with the "Test Py" CI (specifically failing fast for python v2.7). @svenevs You could be using the matrix in the _test_python.yaml_ better to specify `sphinx_ver` and `breathe_ver` with a specific `python_ver`. I would suggest 
```yml
    strategy:
      fail-fast: false
      matrix:
        os: [macos-latest, ubuntu-latest, windows-latest]
        python: [2.7, 3.6, 3.7, 3.8, 3.9]
        include:
          - python: 2.7
            sphinx_ver: '==1.8.5'
            breathe_ver: '==4.12.0'
          - python: 3.6
            sphinx_ver: '==1.8.5'
            breathe_ver: '==4.12.0'
          - python: 3.7
            sphinx_ver: '==2.0.0'
            breathe_ver: '==4.14.0'
          - python: 3.8
            sphinx_ver: '==3.0.0'
            breathe_ver: ''
          - python: 3.9
            sphinx_ver: '==3.0.0'
            breathe_ver: ''
```

**Comment by hidmic** [comment]
@svenevs any chance we can get this moving soon? Some folks and I are willing to share the maintenance load if that helps. FYI @clalancette @nuclearsandwich.

**Comment by clalancette** [comment]
> @svenevs any chance we can get this moving soon? Some folks and I would rather step up and help you maintain this repository and not fork or patch downstream. FYI @clalancette @nuclearsandwich.

For some context; we're some of the core maintainers of https://github.com/ros2 , and we've been developing a tool to generate API documentation on our packages.  The tool is based on doxygen, breathe, exhale, and sphinx, and you can see the implementation at https://github.com/ros-infrastructure/rosdoc2 .  This change, along with #117 and #118 and some fixes in Breathe will allow us to generate the documentation with far fewer errors than before.

As @hidmic said, we'd prefer to help maintain this package rather than carry our own fork.  Please let us know if you are interested in having maintenance help here.  Thanks!

**Comment by svenevs** [comment]
Short answer: yes, sorry, and thank you, help would be most welcome -- you need things faster than I can get to right now. I couldn't rebuild the computer, I'm not able to find the box with the last part. I also can't work on this on my work computer. I will add you as collaborators tomorrow, and you can proceed forward. I can't seem to do it from my phone.

**Comment by clalancette** [comment]
@svenevs Thanks, appreciate the understanding.  I've now accepted the invite.  I'll take a look at some of the open issues in a bit here.

**Comment by svenevs** [comment]
Ok, I was able to fix the things that weren't working afaict.  Please verify that this still does what you want with doxygen pages and let me know.  I just need to update some docs on the required vs optional `rootFileTitle` config now.

https://github.com/svenevs/exhale/pull/114/commits/8eae7ff04331394f5221c2c5c76f7620559ee1ee

changes in `graph.py` there are the main ones to pay attention to, like i was saying nested pages weren't being generated.  they should all be generated now. `self.pages` only includes top level pages, so `generatePageDocuments()` needs to build out the list of nested kids.

@2bndy5 thanks for reiterating the doxygen example i see how the doxygen nesting system works, or at least I think lol.  any page can declare on itself a `subpage` and then doxygen does its thing?  otherwise we're out of scope of what doxygen supports if there is a max level of subpages allowed (see `testing/tests/cpp_nesting.py` and associated `testing/projects/cpp_nesting/page_town_rock*.hpp` files)

hope this is coherent

**Comment by 2bndy5** [comment]
From comment in your last commit about

>  find out why everything breaks in Doxygen 1.9.2

I've come across "unmatched XML tags" in the doxygen output when using v1.9.2 with breathe. I found that the closing XML tags are missing from the doxygen output.

>  any page can declare on itself a subpage and then doxygen does its thing? 

Yes, I'm under the same impression. I have never really used subpages in Doxygen though. And I don't recall the Doxygen manual declaring a limit for nesting pages.

>  otherwise we're out of scope of what doxygen supports if there is a max level of subpages allowed (see testing/tests/cpp_nesting.py and associated testing/projects/cpp_nesting/page_town_rock*.hpp files)

This seems very specific to the PR, and I'm not sure what to make of the _cpp_nesting.py_ code, but the _page_town_rock.hpp_ looks right.

**Comment by svenevs** [comment]
> I've come across "unmatched XML tags" in the doxygen output when using v1.9.2 with breathe. I found that the closing XML tags are missing from the doxygen output.

Oh my! :scream: That's good to know about, definitely not going to do anything about it in this PR then...

Thanks so much for all your work here and on breathe :slightly_smiling_face:

> Yes, I'm under the same impression. I have never really used subpages in Doxygen though. And I don't recall the Doxygen manual declaring a limit for nesting pages.

Haha me neither.  See attached images and zip of html output if helpful below, I think we're more or less on par with doxygen html output.  In some senses, a little better because when there is a use of `\mainpage` nothing actually shows up on `Related Pages` section whereas in exhale it will still be in the page hierarchy :tada:

I'm pretty sure this is the original functionality @hidmic implemented (specifically, root page absorbs `\mainpage` contents and title -- testing framework cheats to achieve it), but it would be good to test on some of the projects you all are after before saying it is good to go.

### with `\mainpage`

with_mainpage

[docs_CPPNestingPages_test_hierarchies_primary_mainpage.zip](https://github.com/svenevs/exhale/files/7614463/docs_CPPNestingPages_test_hierarchies_primary_mainpage.zip)

### without `\mainpage`

without_mainpage

[docs_CPPNestingPages_test_hierarchies_primary_no_mainpage.zip](https://github.com/svenevs/exhale/files/7614465/docs_CPPNestingPages_test_hierarchies_primary_no_mainpage.zip)

---------------------------

ref for myself: #121

**Comment by 2bndy5** [comment]
@svenevs Thanks for the demonstration! I didn't know that "Related Pages" could be used to show a nested hierarchy. That makes the `\subpage` cmd make a lot more sense. Originally, I thought the hierarchy only showed in the TREEVIEW (when enabled in doxygen).

**Comment by svenevs** [comment]
Ok this is rebased and staged to be in the changelog, ready for me to squash merge pending a third party corroborating that they are able to use this branch to build a doxygen project with pages as they expect it to.

I've done all I can for this weekend, it turns out `pip install exhale` is broken right now requiring people to `pip install sphinx` first.  I'd like to merge this and do a release.

There aren't many low hanging fruit left in the issues, so after this release next weekend I'm going to look into the changes you all have been stuffing into breathe -- I want to just depend on the next version to come out (incl. dropping python 2 :upside_down_face:), and ideally find a way to get whatever needs to happen for #106 fixed.  Whether that's fix exhale or update the interface to allow refid lookup on `.. doxygenfunction` or whatever.

**Review by svenevs — APPROVED** [review]

**Comment by 2bndy5** [comment]
Tested this PR with the RF24 project I've been working to convert to sphinx docs...
The pages (which are generated by doxygen from MD files) are re-generated well with exhale. ☑️
The RF24 project does not use `\subpage` cmds, so it isn't a full coverage field test.

**Comment by hidmic** [comment]
Coming back from some time off. Thank you both for pushing this!

**Pull request merged**